### PR TITLE
.Net: Added better formatting for responses from Bing Searches & Ability to use custom Bing Search endpoint.

### PR DESCRIPTION
**Pull request by SOE-YoungS**
### Motivation and Context
Returning just the page snippet from a Bing search results in a lack of context or ability to follow up on a response.

This PR addresses that by adding better formatting for Bing Search responses via the "Plugins.Web" package. 
It also adds a method to use a custom endpoint for Bing Search.

#### Formatting Changes
Previous response format example:
image

New response format example:
image

As can be seen by the examples provided, the new response format contains much more context & provides the user with the URL of the search result so that they can click through and read further.

This can be obtained programmatically later on using regex matching, then could be fed into the "[Search Url Plugin](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Plugins/Plugins.Web/SearchUrlPlugin.cs)", to scrape the page directly, before finally summarizing and returning a complete summary of the page to the user.

#### Endpoint Changes
The ability to use a custom endpoint for Bing Search also enables a path to use an API Management instance as a front end for the Bing Search API. This is required by some end users for scenarios such as enterprise logging and usage counting for cross charge.

The default behaviour for the BingConnector is to target the Bing Search endpoint directly. It is entirely optional to use a custom endpoint. Doing so, does not introduce any other code change requirements to achieve use of Bing Search.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#dev-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by SOE-YoungS** [comment]
Note: Original PR was [#2053](https://github.com/microsoft/semantic-kernel/pull/3357), which I couldn't reopen.

@lemillermicrosoft

**Comment by lemillermicrosoft** [comment]
> Note: Original PR was [#2053](https://github.com/microsoft/semantic-kernel/pull/3357), which I couldn't reopen.
> 
> @lemillermicrosoft

@dmytrostruk FYI

**Comment by dmytrostruk** [comment]
> Note: Original PR was [#2053](https://github.com/microsoft/semantic-kernel/pull/3357), which I couldn't reopen.
> 
> @lemillermicrosoft

@SOE-YoungS Thank you for your contribution!
Do you want to re-open previous PR or work in scope of this one?

**Comment by SOE-YoungS** [comment]
@lemillermicrosoft,  @dmytrostruk, Any chance this PR could be reviewed please?

We're currently blocked from updating the beta5 to the beta6 Nuget packages, due to changes in Microsoft.SemanticKernel.Abstractions that won't be resolved for us until this PR is merged and the changes introduce are released.

**Review by lemillermicrosoft — COMMENTED** [review]
A few comments, overall LGTM. CC @markwallace-microsoft to make sure another appropriate set of eyes are assigned.

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:36]: I didn't modify this, it was like this in the previous commits too.

Happy to update it, but I'm not sure what the issue is here?

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs]: Resolved in 615cbc4a86763aade2b3e21ca8f2ba27ffcdd9a3

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:58]: Resolved in 615cbc4a86763aade2b3e21ca8f2ba27ffcdd9a3

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs]: Resolved in 615cbc4a86763aade2b3e21ca8f2ba27ffcdd9a3

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (lemillermicrosoft) [inline dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:32]: nit: Add some details.

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (lemillermicrosoft) [inline dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:58]: You could use `Uri? uri` in the constructor and this line be `this._uri = uri ?? new Uri(DefaultUri)`

**Review by lemillermicrosoft — APPROVED** [review]
Overall LGTM. A few nit picks. Would love to integrate with these changes, too!

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:58]: @lemillermicrosoft a Uri? cannot be instantiated as a compile time constant. 

As a result this use would (does) throw errors, hence the approach I took as seen in commits for this PR.

image

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs]: Resolved in https://github.com/microsoft/semantic-kernel/pull/3357/commits/9be54798d946929a1e09881aa6c31315cffb5114

File: dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs
Comment (lemillermicrosoft) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:58]: The default in the constructor would be null instead.

Comment (lemillermicrosoft) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs:58]:
```
public BingConnector(string apiKey, Uri? uri = null, ILoggerFactory? loggerFactory = null)
```

Comment (SOE-YoungS) [reply dotnet/src/Plugins/Plugins.Web/Bing/BingConnector.cs]: Ah Ok, from the PR previously, it had been mentioned that the constructor should reveal the DefaultUri being used (when hovered over).

Happy to make this change though (it's in https://github.com/microsoft/semantic-kernel/pull/3357/commits/485ca374c1af20d824b9f45f142395ca1e4be445)

**Comment by SOE-YoungS** [comment]
@dmytrostruk @markwallace-microsoft 

Any chance we can get this reviewed for second approval please?

@lemillermicrosoft

**Review by markwallace-microsoft — APPROVED** [review]

**Comment by markwallace-microsoft** [comment]
@SOE-YoungS Can you please take a look at the build failures associated with this PR?

**Comment by markwallace-microsoft** [comment]
@SOE-YoungS I'm going to close this for now but please feel free to re-open when the build failures are resolved

**Comment by SOE-YoungS** [comment]
@markwallace-microsoft I've brought the code for this PR up to date with Microsoft/main. 
However, I don't seem to be able to re-open this PR. 

How should I proceed?

**Comment by SOE-YoungS** [comment]
Just to update: I've created another draft PR #5673.

Once all checks are passing, I'll mark it as ready for review.

**Pull request closed without merging**
